### PR TITLE
Discord Verification - Credential Issue Backend

### PR DESCRIPTION
**Pull request by kevinz917**
- Created the credential issue flow in Rust for Discord. 
- This does not include the UI changes that are the second part of this effort to integrate Discord flow.

### CODE REVIEW

File: worker/worker/worker.js [hunk 1]
```diff
@@ -147,50 +147,89 @@ function handleOptions(request) {
...
+  try {
+    const { witness_discord } = wasm_bindgen;
+    const { searchParams } = new URL(request.url);
+
+    const pk = decodeURI(searchParams.get("pk"));
+    const message_id = searchParams.get("messageId");
+    const channel_id = searchParams.get("channelId");
+    const discord_handle = searchParams.get("discordHandle");
+
+    await wasm_bindgen(wasm); // What is this line for?
```
Comment (sbihel) [inline worker/worker/worker.js:208]: Interfaces with WASM can be a bit messy, at least from what is documented, and CF adds more abstractions. I think it's fair to say that this simply loads the module (and makes the interface actually callable).

File: worker/src/lib.rs [hunk 2]
```diff
@@ -135,6 +134,93 @@ pub async fn witness_tweet(
...
+        );
+
+        evidence_map.insert(
+            "channelId".to_string(),
+            serde_json::Value::String(discord_res.channel_id),
+        );
+        evidence_map.insert(
+            "messageId".to_string(),
+            serde_json::Value::String(discord_res.id),
+        );
```
Comment (sbihel) [inline worker/src/lib.rs:215]: You can initialise the map from an array if you want to make it less verbose https://doc.rust-lang.org/std/collections/struct.HashMap.html#examples

File: worker/src/lib.rs [hunk 3]
```diff
@@ -135,6 +134,93 @@ pub async fn witness_tweet(
...
+                .await
+        );
+        let formatted_discord_handle = format!(
+            "{}#{}",
+            discord_res.author.username, discord_res.author.discriminator
+        );
+        let mut vc = jserr!(discord::build_discord_vc(&pk, &formatted_discord_handle));
+
+        // Check for matching handles
+        if discord_handle != formatted_discord_handle {
```
Comment (sbihel) [inline worker/src/lib.rs:180]: If Discord handles aren't case sensitive, you should compare them in their lowercased form

File: worker/src/lib.rs [hunk 4]
```diff
@@ -135,6 +134,93 @@ pub async fn witness_tweet(
...
+pub async fn witness_discord(
+    secret_key_jwk: String,
+    public_key_tezos: String,
+    discord_authorization_key: String,
+    discord_handle: String,
+    channel_id: String,
+    message_id: String,
+) -> Promise {
+    use log::Level;
+    console_log::init_with_level(Level::Trace).expect("error initializing log");
```
Comment (sbihel) [inline worker/src/lib.rs:164]: Could you factor out the logging initialisation and put the level to Error so we don't have unimportant logs in production? (I don't even know if there's a way to get logs from live workers)

File: worker/src/discord.rs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        .headers(headers)
+        .send()
+        .await?
+        .json()
+        .await?;
+
+    Ok(res)
+}
+
+pub fn build_discord_vc(pk: &JWK, discord_handle: &str) -> Result<Credential> {
```
Comment (sbihel) [inline worker/src/discord.rs:73]: Could you put the Twitter one in its file as well?

File: worker/worker/worker.js [hunk 6]
```diff
@@ -147,50 +147,89 @@ function handleOptions(request) {
...
+    const discordVc = await witness_discord(
+      TZPROFILES_ME_PRIVATE_KEY,
+      pk,
+      DISCORD_AUTHORIZATION_TOKEN,
+      discord_handle,
+      channel_id,
+      message_id
+    );
+
+    console.log(discordVc);
```
Comment (sbihel) [inline worker/worker/worker.js:218]: Should probably remove this log line and manage all logs within Rust.

File: worker/src/discord.rs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+use serde::Deserialize;
+use serde_json::json;
+use ssi::{
+    blakesig::hash_public_key,
+    jwk::{Algorithm, JWK},
+    jws::verify_bytes,
+    one_or_many::OneOrMany,
+    tzkey::jwk_from_tezos_key,
+    vc::{Credential, Evidence, LinkedDataProofOptions},
+};
```
Comment (sbihel) [inline worker/src/discord.rs:13]: I think some of those imports aren't used.

**Review by sbihel — CHANGES_REQUESTED** [review]

File: worker/src/discord.rs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                  }
+              }
+          }
+      ],
+      "issuanceDate": Utc::now().to_rfc3339_opts(SecondsFormat::Millis, true),
+      "id": format!("urn:uuid:{}", Uuid::new_v4().to_string()),
+      "type": ["VerifiableCredential", "DiscordVerification"],
+      "credentialSubject": {
+          "id": format!("did:pkh:tz:{}", &hash_public_key(pk)?),
+          "sameAs": "urn:discord.com:".to_string() + discord_handle
```
Comment (sbihel) [inline worker/src/discord.rs:82]: Does Discord have a link for a profile, to avoid inventing a new one?

File: worker/src/lib.rs (hunk 3 above)
Comment (kevinz917) [reply worker/src/lib.rs:180]: I believe they are case sensitive

File: worker/src/discord.rs (hunk 5 above)
Comment (kevinz917) [reply worker/src/discord.rs:55]: Refactored

File: worker/src/discord.rs (hunk 8 above)
Comment (kevinz917) [reply worker/src/discord.rs:83]: I don't think Discord has public display profile links :(.

**Pull request merged**